src/subtitles-octopus.js

```diff
@@ -404,14 +410,50 @@ var SubtitlesOctopus = function (options) {
         }
     };
 
+    function _computeCanvasSize(width, height) {
+        var hardHeightLimit = Math.max(self.hardHeightLimit || height, self.softHeightLimit);
+
+        if (self.prescaleTradeoff === null) {
+            if (height > hardHeightLimit) {
+                width = width * hardHeightLimit / height;
+                height = hardHeightLimit;
+            }
+        } else if (self.prescaleTradeoff > 1) {
+            if (height * self.prescaleTradeoff <= self.softHeightLimit) {
+                width *= self.prescaleTradeoff;
+                height *= self.prescaleTradeoff;
+            } else if (height < self.softHeightLimit) {
+                width = width * self.softHeightLimit / height;
+                height = self.softHeightLimit;
+            } else if (height > hardHeightLimit) {
+                width = width * hardHeightLimit / height;
+                height = hardHeightLimit;
+            }
+        } else if (height > self.softHeightLimit) {
+            if (height * self.prescaleTradeoff <= self.softHeightLimit) {
+                width = width * self.softHeightLimit / height;
+                height = self.softHeightLimit;
+            } else if (height * self.prescaleTradeoff <= hardHeightLimit) {
+                width *= self.prescaleTradeoff;
+                height *= self.prescaleTradeoff;
+            } else {
+                width = width * hardHeightLimit / height;
+                height = hardHeightLimit;
+            }
+        }
```

Comment:
In this case, it would probably be better if you port these features.

Reply:
> In this case, it would probably be better if you port these features.

Should I push the amended and redistributed commits to this branch (same PR) or would you prefer a new PR being opened for this?
If you want to keep splitting the jellyfin parts into several standalone series, then for smaller series which don't require familiarity with JS and its worker-model *(so not `renderAhead` itself)*, I can *maybe* do the squashing and splitting if you absolutely do not want to do that. However, each series must only deal with one concise topic at a time.

Reply:
> Should I push the amended and redistributed commits to this branch (same PR) or would you prefer a new PR being opened for this?

I think "new PR" since it will be entirely your changes - you literally rewrote the feature.

> If you want to keep splitting the jellyfin parts into several standalone series, then for smaller series which don't require familiarity with JS and its worker-model (so eg not renderAhead itself), I can maybe do the squashing and splitting if you absolutely do not want to do that.

The main problem why you and me will never agree is that your team want to have the perfect contribution all at once, and I want to preserve commit history and original contributions (or at least not ruin them too much). I also like the idea of keeping the branch as clean as possible, but this doesn't work well in case of porting - you lose details about contributions (commit message doesn't point the exact lines).

As I said, I can squash some minor commits to where the relevant code appeared first time or an error was introduced, but only if the author is the same. I already did this with that `Bump height` commit.

My porting scheme: `original feature, then fixes/improvements/etc.`, that way I'm sure I won't break something that already exists (in the feature), and I can review the history in case further "improvements" break it.

I treat the whole branch (PR) as a feature. This gives me more flexibility in the content of the branch, focusing only on the result of the merge.

When bisecting, in any case, you need to localize the bad branch (merge commit) by traversing only the parent (`master`) branch. Then you bisect the branch. I agree, how quickly you find the error will depend on the quality of the branch. But what do you want to bisect in small features like this?
By squashing commits, you remove the details, and you don't even need to look into the branch - merge commit is the final point.

> However, each series must only deal with one concise topic at a time and the commits themselves should also not contain a bunch of unrelated changes.

I totally agree with you in case of a brand-new features by a single author. I tend to open a separate PR for unrelated changes, as I did for the buffer refactor to free up the `renderAhead` branch.

Current features "live in harmony" with each other, sharing `_computeCanvasSize` function. Splitting simply won't provide any significant benefit.

I thank you for your detailed posts and help. I see you want the best for the project, but wouldn't it be better to merge the original features first (with or without patches) and then refactor them as you want?

Currently, there are 4 branches:
- Refactor of buffer
- Canvas size limit (hard limit + prescale)
- renderAhead (renderAhead + dropAnimation)
- Smart blend

Reply:
> I think "new PR" since it will be entirely your changes - you literally rewrote the feature.

Will do so then in a couple of days or so.

> but wouldn't it be better to merge the original features first (with or without patches) and then refactor them as you want?

Not only is this bad for bisecting and introduces (semi-)broken states into the repository and its history; it is also a hindrance for review and will lead to additional unanticipated  bugs slipping through. See for example how I found issues with values `<= 0` and could simplify the logic in this PR, and found the (pre-existing) signed-overflow/out-of-bounds-write issues in the buffer PR only after they've been split out. Previously I missed them because I was too distracted by having to keep all future changes and known bugs to ignore in mind.
Unlike this PR which only changes 49 lines in total, for `renderAhead` splitting alone will not be sufficient to enable proper review and at the same time, the size and complexity of it makes it the most likely to still have bugs lurking.
Small, atomic well-documented commits enable proper review **and** proper bisecting, which reduces the bugs that are merged at all and makes it easier to deal with the remaining bugs which slip through. In the current state of the renderAhead-branch the individual commits are neither atomic, small (within the bounds of what's possible) nor well-documented.

Reply:
> Not only is this bad for bisecting and introduces (semi-)broken states into the repository and its history;

When using merge commits, the `master` branch will stay unbroken.
Actually, it depends on what is considered "broken". You anyway might miss something or even make a bug. But you can fix the bug or improve the code later.

> Previously I missed them because I was too distracted by having to keep all future changes and known bugs to ignore in mind.

That's because you are reviewing per commit.
Again, you could fix them later.